Cpp_test/lib/ipc-shm.h

```diff
@@ -0,0 +1,45 @@
+#ifndef IPC_SHM_H
+#define IPC_SHM_H
+
+#include <iostream>
+#include <pthread.h>
+#include <unistd.h> 
+#include <cstring>
+#include <fcntl.h>
+#include <sys/mman.h>
+#include <vector>
+
+struct ipc_shm_header_t
+{
+	pthread_cond_t cond;
+	pthread_mutex_t mutex;
+	std::streamsize data_size;
+	size_t data_version;
+	std::streamsize shared_mem_size;
+	volatile bool is_init;
+	volatile bool is_read;
+	char data_ap[1024];
+};
```

Comment:
You can't be sure that the struct will be ordered in the same way for the two programs (depending of the compiler and the compiler options).

I think you have two options:
* structure your struct with some offset defined, and use memcpy with the offset, like:
```c++
struct field{uint32_t offset; char[8] name;};
struct ipc_shm_header_t
{
    uint32 preamble_len;
    field preamble [NB_FIELDS];
    pthread_cond_t cond;
    pthread_mutex_t mutex;
    std::streamsize data_size;
    size_t data_version;
    std::streamsize shared_mem_size;
    volatile bool is_init;
    volatile bool is_read;
    char data_ap[1024];
};
```
You can compare the `preamble_len` and the `preamble` to be __quite__ sure that the struct is ordered in the same way.

* Use a struct of pointer and set the offset manually after mmap the shm.

Reply:
I think the struct should be ordered. Can you give me an example or information about this thing?

Reply:
Basically, it should be ordered correctly, but you don't have any warranty for this.


You can build one program like usual, and the other with `-fpack-struct[=n]`

Reply:
As you said, basically it works in order. And in my Bazel build, there are no flags that can make it work in an inappropriate way. So, we do not need to add the -fpack-struct[=n] to complicate it.

Reply:
Yes, but the objective of the academy is to learn about how coding, and getting good habits. In a real project, this won't be ok to implement this, it's not safe.